### PR TITLE
fix(custom-resources): custom handler status not being used by framework

### LINKED ISSUE #22696: custom-resources: Does not conform to CFN's Custom Resource Response Specification
### Describe the bug

The current implementation of custom-resources provider-framework does not correctly reflect what handlers are expect to respond with based on [CFN's own resource specification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/crpg-ref-responses.html). 

Specifically the `Status` parameter of the response object is being ignored by the provider-framework and always returning success even when a Lambda properly returned a formatted failure, this has lead to custom resources that we've created incorrectly returning CREATE_COMPLETE even though there was an error. It's noted in the documentation for the custom-resources provider that `onEvent` handlers are expected to throw in the case of an error however CFN does not expect them to throw and instead expects them to either throw or more ideally return with a formatted error message.

### Expected Behavior

The provider framework should expect that a a valid`Status` of either `'SUCCESS' | 'FAILED'` is return from the customer provider `onEvent` handler. The indication of failure in the case of exceptions being thrown should still be supported to properly bubble up errors to customers that are un-handled.

### Current Behavior

Provider ignores the status returned by an `onEvent` handler and instead just returns `SUCCESS` even in the case of a `FAILED` response. This leads to state bugs in CFN stack resources that are CustomResources back by this framework with a handler that is properly configured per CFN's specification.

### Reproduction Steps

```ts
// Example onEvent handler

const handler = (event) => {
  const response = {
    Status: 'FAILED',
    Reason: "Some valid reason to fail",
    StackId: event.StackId,
    RequestId: event.RequestId,
    response.Data = {
      Response: null
    }
  }

  return response;
}

const provider = new customResource.Provider(this, 'Provider' {
  onEventHandler: handler
});

// Implement some custom resource to manager here.
```

### Possible Solution

There's a few potential solutions to this that are simple,

1. Always expect a valid response to contain a Status outside of exceptions (BREAKING) [Recommended] 
2. Keep existing success behavior unless a Status has been submitted (Non-Breaking maybe?)

### Additional Information/Context

For context anybody using this framework at the moment should be impacted, unless they're using an anti-pattern of just not handling any exceptions and throwing.

### CDK CLI Version

2.49.0

### Framework Version

_No response_

### Node.js Version

v18.4.0

### OS

Mac/Linux

### Language

Typescript

### Language Version

4.8

### Other information

_No response_

### PR DESCRIPTION
**Pull request by ianmarmour**
### Describe the bug

The current implementation of custom-resources provider-framework does not correctly reflect what handlers are expect to respond with based on [CFN's own resource specification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/crpg-ref-responses.html).

Specifically the Status parameter of the response object is being ignored by the provider-framework and always returning success even when a Lambda properly returned a formatted failure. This has shown up as a bug in custom resources that we've created incorrectly returning CREATE_COMPLETE even though there was an error and a `FAILED` response from our Lambda. It's noted in the documentation for the custom-resources provider that `onEvent` handlers are expected to throw in the case of an error however CFN does not expect them to throw and instead expects them to either throw or more ideally return with a formatted error message.

### Root Cause

Original implementation ignored the schema of CFN and instead always succeeded.

### Solution

Optionally allow for Status to be utilized if it's provided by the response from a custom resource handler.

fixes #22696 
----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -2602,7 +2602,7 @@ available-typed-arrays@^1.0.5:
   resolved "https://registry.npmjs.org/available-typed-arrays/-/available-typed-arrays-1.0.5.tgz#92f95616501069d07d10edb2fc37d3e1c65123b7"
   integrity sha512-DMD0KiN46eipeziST1LPP/STfDU0sufISXmjSgvVsoU2tqxctQeASejWcfNtxYKqETM1UxQ8sp2OrSBWpHY6sw==
 
-aws-cdk-lib@^2.47.0:
+aws-cdk-lib@2.47.0:
```
Comment (pflorek) [inline yarn.lock:2605]:
```suggestion
aws-cdk-lib@^2.47.0:
```

Revert this

**Review by pflorek — CHANGES_REQUESTED** [review]
@ianmarmour nice Feature

Please implement both `OnEvent` and `IsComplete`.


Please update the response types
https://github.com/aws/aws-cdk/blob/fadbfc1eb07f4f2daecfe623812fee029c81e31a/packages/%40aws-cdk/custom-resources/lib/provider-framework/types.d.ts#L53

File: yarn.lock (hunk 1 above)
Comment (TheRealAmazonKendra) [inline yarn.lock:2605]: No changes to yarn.lock, please.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Thank you for your contribution. We need a succeeding build and for all PR linter status checks to pass before we can provide a meaningful review.

**Comment by mrgrain** [comment]
Hi @ianmarmour Before you go off and invest much more work on this, we need to check if this is a possible change. I am confirming the original design with the team, so please bear with me and I might be wrong.

--- 

1) To go with lambda exit "status" was a conscious design. See the big yellow box on [this documentation page](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources-readme.html):
>N.B.: if you use the provider framework in this module you will write AWS Lambda Functions that look a lot like, but aren't exactly the same as the Lambda Functions you would write if you wrote CloudFormation Custom Resources directly, without this framework.
>
>Specifically, to report success or failure, have your Lambda Function exit in the right way: return data for success, or throw an exception for failure. Do not post the success or failure of your custom resource to an HTTPS URL as the CloudFormation documentation tells you to do.

2) You are arguing that the CFN specification stipulates a certain response format for the Lambda. However the page you've linked states (highlight is mine):
> The following are properties that the custom resource provider includes **_when it sends the JSON file to the presigned URL_**.

Again, this describes a different interface.

3) The proposed change would also break any Customers that currently return a `Status` other than `SUCCESS`. This might not affect many, but it would be a breaking change that at least has to go behind a feature flag.


---

If you'd rather use the original CloudFormation Custom Resource framework, that's also available: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.CustomResource.html

**Review by mrgrain — CHANGES_REQUESTED** [review]
Until resolved, see above.

**Comment by ianmarmour** [comment]
@mrgrain

> To go with lambda exit "status" was conscious design. See the big yellow box on [this documentation page](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources-readme.html):

I very much missed that when implementing our custom resources 😆. 

> The following are properties that the custom resource provider includes when it sends the JSON file to the presigned URL.

I think my confusion here mostly stemmed from the fact that the provider-framework is actually invoking a lambda and that lambda isn't really a custom resource handler it's a provider-framework handler with it's own behavior specifically needed to conform with this framework (the framework provides some things that are very nice to have). When I was originally working on our code that leveraged this I kind of assumed the framework was much less opinionated then it is in practice.

> The proposed change would also break any Customers that currently return a Status other than SUCCESS. This might not affect many, but it would be a breaking change that at least has to go behind a feature flag.

I would wonder in what case returning FAILURE and not failing wouldn't be an anti-pattern? I agree that it would break existing customers however I can't imagine a time when a customer is intentionally emitting a failure but wants the resource to be added to the stack successfully (though I suppose there could be a valid use case there). If that use case exists it seems like something that the CFN team should support directly? 

I'm happy to close this issue for now though and update my personal code to throw rather then return a status.

**Comment by mrgrain** [comment]
> I very much missed that when implementing our custom resources 😆.
> I think my confusion here mostly stemmed from the fact that the provider-framework is actually invoking a lambda and that lambda isn't really a custom resource handler it's a provider-framework handler with it's own behavior specifically needed to conform with this framework (the framework provides some things that are very nice to have). When I was originally working on our code that leveraged this I kind of assumed the framework was much less opinionated then it is in practice.

I'm glad we could clear this up! 👍🏻 It seems like we could still improve how we highlight the difference between the two. If you've got any suggestions on what we could improve documentation wise, they would be very welcome.

> I would wonder in what case returning FAILURE and not failing wouldn't be an anti-pattern? I agree that it would break existing customers however I can't imagine a time when a customer is intentionally emitting a failure but wants the resource to be added to the stack successfully (though I suppose there could be a valid use case there). If that use case exists it seems like something that the CFN team should support directly?

Agreed, it's very unlikely. I wasn't actually thinking about returning `FAILURE` specifically, but literally anything else. Like `OK` or just passing through the response from a library and ending up with an arbitrary status (e.g. HTTP status codes).

> I'm happy to close this issue for now though and update my personal code to throw rather then return a status.
👍🏻

**Comment by Lewenhaupt** [comment]
I'm using [aws-cloudformation-custom-resource](https://www.npmjs.com/package/aws-cloudformation-custom-resource) and it returns FAILED when failing but I just discovered this when Cloudformation thought everything was working but all our db migrations were missing in a new stack. Sure one could use the old Provider, but that (for some reason) decided you can't add an existing lambda to it, only create a new one, which doesn't really fit what we're doing.

I think it's very strange this doesn't adhere to the custom resource framework of CFN.

**Pull request closed without merging**